src/gamepad/linux/linux-raw-gamepad.vala

```diff
@@ -0,0 +1,150 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.LinuxRawGamepad : Object, RawGamepad {
+	private int fd;
+	private uint? event_source_id;
+	private Libevdev.Evdev device;
+
+	private uint8 key_map[Linux.Input.KEY_MAX];
+	private uint8 abs_map[Linux.Input.ABS_MAX];
+	private Linux.Input.AbsInfo abs_info[Linux.Input.ABS_MAX];
+	private string identifier;
+
+	private string _name;
+	public string name { get { return _name; } }
+
+	private string _guid;
+	public string guid { get { return _guid; } }
+
+	private uint8 _axes_number = 0;
+	public uint8 axes_number { get { return _axes_number; } }
+
+	private uint8 _buttons_number = 0;
+	public uint8 buttons_number { get { return _buttons_number; } }
+
+	private int _dpads_number = -1;
+	public uint8 dpads_number {
+		get {
+			if (_dpads_number != -1)
+				return (uint8) _dpads_number;
+
+			_dpads_number = 0;
+			for (var i = Linux.Input.ABS_HAT0X; i <= Linux.Input.ABS_HAT3Y; i += 2) {
+				if (device.has_event_code (Linux.Input.EV_ABS, i) ||
+				    device.has_event_code (Linux.Input.EV_ABS, i + 1)) {
+					var absinfo = device.get_abs_info (i);
+					if (absinfo == null)
+						continue;
+
+					_dpads_number++;
+				}
+			}
+
+			return (uint8) _dpads_number;
+		}
+	}
+
+	public LinuxRawGamepad (string file_name) throws FileError {
+		identifier = file_name;
+		fd = Posix.open (file_name, Posix.O_RDONLY | Posix.O_NONBLOCK);
+
+		if (fd < 0)
+			throw new FileError.FAILED (@"Unable to open file $file_name: $(Posix.strerror (Posix.errno))");
+
+		device = new Libevdev.Evdev ();
+		if (device.set_fd (fd) < 0)
+			throw new FileError.FAILED (@"Evdev is unable to open $file_name: $(Posix.strerror (Posix.errno))");
+
+		_name = device.name;
+		_guid = LinuxGuidHelpers.from_dev (device);
+
+		// Poll the events in the default main loop
+		var channel = new IOChannel.unix_new (fd);
+		event_source_id = channel.add_watch (IOCondition.IN, poll_events);
+
+		// Initialize dpads, buttons and axes
+		for (var i = Linux.Input.BTN_JOYSTICK; i < Linux.Input.KEY_MAX; i++) {
+			if (device.has_event_code (Linux.Input.EV_KEY, i)) {
+				key_map[i - Linux.Input.BTN_MISC] = _buttons_number;
+				_buttons_number++;
+			}
+		}
+		for (var i = Linux.Input.BTN_MISC; i < Linux.Input.BTN_JOYSTICK; i++) {
+			if (device.has_event_code (Linux.Input.EV_KEY, i)) {
+				key_map[i - Linux.Input.BTN_MISC] = _buttons_number;
+				_buttons_number++;
+			}
+		}
+
+
+		// Get info about axes
+		for (var i = 0; i < Linux.Input.ABS_MAX; i++) {
+			/* Skip dpads */
+			if (i == Linux.Input.ABS_HAT0X) {
+				i = Linux.Input.ABS_HAT3Y;
+				continue;
```

Comment:
https://github.com/Kekun/gnome-games/blob/master/CodingStyle.txt. You know there is a code style which states there should be a new line before return, throw, break, **etc.** That includes **continue**. No need to be rude.